synapseclient/core/download/download_async.py

```diff
@@ -326,169 +266,286 @@ class _MultithreadedDownloader:
-        # 1. scheduling any additional part downloads as previous parts are completed
-        # 2. writing any completed parts out to the local disk
-        # 3. waiting for additional parts to complete
-        pending_futures = set()
-        completed_futures = set()
-        try:
-            while True:
-                submitted_futures = self._submit_chunks(
-                    url_provider,
-                    chunk_range_generator,
-                    pending_futures,
-                )
-
-                self._write_chunks(request, completed_futures, transfer_status)
+        file_size = await _get_file_size_wrapper(
+            syn=self._syn, url=url_info.url, debug=self._download_request.debug
+        )
+        self._create_progress_bar(file_size=file_size)
+        self._prep_file()
 
-                # once there is nothing else pending we are done with the file download
-                pending_futures = pending_futures.union(submitted_futures)
-                if not pending_futures:
-                    break
+        # Create AsyncIO tasks to download each of the parts according to chunk size
+        download_tasks = self._generate_stream_and_write_chunk_tasks(
+            url_provider=url_provider,
+            chunk_range_generator=_generate_chunk_ranges(file_size),
+        )
+        await self._execute_download_tasks(download_tasks=download_tasks)
 
-                completed_futures, pending_futures = concurrent.futures.wait(
-                    pending_futures, return_when=concurrent.futures.FIRST_COMPLETED
-                )
+    async def _execute_download_tasks(self, download_tasks: Set[asyncio.Task]) -> None:
+        """Handle the execution of the download tasks.
 
-                self._check_for_errors(request, completed_futures)
+        Arguments:
+            request: A DownloadRequest object specifying what Synapse file to download.
+            download_tasks: A set of asyncio tasks to download the file in chunks.
 
-        except BaseException:
-            # on any exception (e.g. KeyboardInterrupt), attempt to cancel any pending futures.
-            # if they are already running this won't have any effect though
-            for future in pending_futures:
-                future.cancel()
+        Returns:
+            None
+        """
+        cause = None
+        while download_tasks:
+            done_tasks, pending_tasks = await asyncio.wait(
+                download_tasks, return_when=asyncio.FIRST_COMPLETED
+            )
+            download_tasks = pending_tasks
+            for completed_task in done_tasks:
+                try:
+                    start_bytes, end_bytes = completed_task.result()
+                    del completed_task
+                    self._syn._parts_transfered_counter += 1
+
+                    # Garbage collect every 100 iterations
+                    if self._syn._parts_transfered_counter % 100 == 0:
```

Comment:
Would there ever be a case where the total # of parts the file has been split into never exceeds 100 and therefore garbage collection never happens?

Reply:
Yes it would occur if the file is not 800MIB or larger. But in this case it'll get cleaned up during normal GC runs.

It isn't an issue in this case as other parts of the code have enough references that get created/released that'll cause GC to kick off it's normal garbage culling